sdk/src/main/java/io/opentelemetry/sdk/trace/export/SimpleSpanProcessor.java

```diff
@@ -71,15 +74,30 @@ public boolean isStartRequired() {
   }
 
   @Override
+  @SuppressWarnings("BooleanParameter")
   public void onEnd(ReadableSpan span) {
-    if (sampled && !span.getSpanContext().getTraceFlags().isSampled()) {
-      return;
-    }
-    try {
-      List<SpanData> spans = Collections.singletonList(span.toSpanData());
-      spanExporter.export(spans);
-    } catch (Throwable e) {
-      logger.log(Level.WARNING, "Exception thrown by the export.", e);
+    if (exportAvailable.compareAndSet(true, false)) {
```

Comment:
For `SimpleSpanProcessor` I think I'd expect it to be simple and just call export in all cases (if there is a backup in the exporter, that is a reason to use a more complicated span processor like batch span processor).

Reply:
That would then require exporters to become re-entrant. Our assumption at present is that only one export is called at a time (which is probably something we should document - it can simplify the authoring of exporters - wdyt?).

Reply:
I'm not sure what you mean by "re-entrant" here. Do you just mean having more than one export running at a time, or are you actuallyreferring to sending the same spans more than once and having the exporter be able to handle de-duping them?

Reply:
Meanwhile, I've added some API docs to clearly state the position on re-entrancy.

Reply:
> I'm not sure what you mean by "re-entrant" here. Do you just mean having more than one export running at a time, or are you actuallyreferring to sending the same spans more than once and having the exporter be able to handle de-duping them?

I was referring to the number of `export` or `flush` calls in flight at one time, which is what this code is associated with. We only want one of each of these calls in flight. In fact, we shouldn't really call `export` and `flush` at the same time either. Keeps the programming model simpler for exporters.

Reply:
BTW I'm not seeing that the SDK ever calls on `flush`. I'm now wondering about its utility. :-)

Reply:
The idea with `flush` wasn't for the SDK itself to call it, but for (for example), a lambda to be able to call `flush` explicitly, so that spans wouldn't be left in a queue, unsent, when the lambda shuts down.

Reply:
Thanks for the clarification. Should we be marking it as `protected` then, to clarify the intent?

Reply:
> We only want one of each of these calls in flight.

I dunno about this - since an exporter is a function that takes spans and pushes them to a sink, I'd be surprised that it would generally matter whether there is one or more in flight. An exporter that depended on being called only once sounds like it's more complicated than it should be. I think right now, if `SimpleSpanProcessor` is used and two spans are ended at the same time, maybe as a result of being parallel client RPCs, one of them is guaranteed to be dropped right? I don't think that's the intention of the `SimpleSpanProcessor`.

Reply:
Ah - very good point. I'll document them as being re-entrable.

Reply:
>  Should we be marking it as protected then, to clarify the intent?

By lambda, I think @jkwatson meant AWS Lambda, not Java lambda, or other Google Cloud Functions type of serverless environment where background threads get suspended as soon as a response is returned. So `flush` needs to stay `public` to be called from the app.

Reply:
Got it. Thanks.

Reply:
I've just made another commit that now clarifies that an `export` or `flush` can be called simultaneously. This thinking has really cleared up the code as I was also able to remove the barrier that enforced one `export` at a time. All tests appear to be working locally for me now.

Reply:
@anuraaga I'm worried that the `OtlpGrpc` and other exporters have no appearance of backpressure in place. I'm presuming that when there's a problem trying to connect, many requests will pile up (this could happen prior to this PR).

The exporters should generally handle `CompletableResultCode.whenCompleted`, test for failure and then do what is necessary to cancel any in-flight request. Additionally, an exporter might decide to avoid another export if there's one in-flight and also buffer a small amount.

With Akka-streams, writing an exporter is quite straightforward. I'd create a bounded queue and drop new elements given any back-pressure. This can mostly be achieved with a `Source.queue` call. We don't have Akka-streams here though. :-)

Any thoughts on this?

Reply:
I think this sort of backpressure is handled by the batch processor which drops when the queue is full.

The simplespanprocessor doesn't but it's why we recommend batch for most apps, simple is more specialized that requires some thought to use.

This seems ok to me but what do you think?

Reply:
I think we should document `SimpleSpanProcessor` with a cautionary note. I shall do this now.

Reply:
Having said that, the `IntervalMetricReader` has no such back-pressure mechanism.

Reply:
So, should I re-instate the ability to only call `export` one at a time for the `IntervalMetricReader` and `BatchSpanProcessor`, and have them drop any new telemetry in the case where an `export` is in-flight? We can leave the `SimpleSpanProcessor` as-is though i.e. without the "in-flight" check. The `SimpleSpanProcessor` will be perfect for those more sophisticated exporters, such as what can be written with Akka-streams.

Reply:
This SGTM

Reply:
I've now pushed a commit with updated documentation for the exporters, and the reinstatement of backpressure handling for the `IntervalMetricReader` and `BatchSpanProcessor`.

BTW Is there an equivalent of `SimpleSpanProcessor` for metrics?

Reply:
Metrics have a somewhat different pipeline structure. The closest equivalent is probably the `IntervalMetricReader` class for push-based exporters. Pull-based, of course, manage all of that themselves.